lms/util/h.py

```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
+
+"""Utility functions for working with the h API."""
+
+from __future__ import unicode_literals
+
+import random
+import string
+
+
+ALLOWED_CHARS = string.ascii_letters + string.digits + "._"
+"""The characters that are allowed in h usernames."""
+
+
+def generate_username(request_params):
+    """Return an h username generated from the given LTI launch request params."""
+
+    full_name = request_params.get("lis_person_name_full") or ""
+    user_id = request_params.get("user_id") or ""
+
+    # Take up to 22 valid characters from full_name.
+    # (h usernames can be up to 30 chars long, and we will be appending 8 chars
+    # from user_id below, leaving us up to 22 chars from full_name).
+    sanitized_full_name = _sanitize_string(full_name, limit=22) or "Anonymous"
+
+    sanitized_user_id = _pad_with_random_characters(
+        _sanitize_string(user_id, limit=8), 8
+    )
+
+    return sanitized_full_name + sanitized_user_id
+
+
+def _sanitize_string(unsanitized_string, limit):
+    """Return ``unsanitized_string`` with invalid chars removed.
+
+    Remove any characters not valid in an h username, and also truncate the
+    resulting string of valid characters to ``limit`` characters long.
+    """
+    sanitized_string = ""
+    for char in unsanitized_string:
+        if char in ALLOWED_CHARS:
+            sanitized_string = sanitized_string + char
+        if len(sanitized_string) >= limit:
+            break
+    return sanitized_string
+
+
+def _pad_with_random_characters(s, length):
```

Comment:
Do you think it's worth breaking this out as a function? Do you foresee reuse? It may be easier to follow the username-building logic if this snippet were part of the `generate_username` function.

Reply:
I like breaking things out into separate functions just for clean coding and readability, because I think the "top level" algorithm reads better when it can be written using higher level nouns and verbs and with all the detail noise removed. Defining certain values as `CONSTANTS` at the top of the file does the same thing but with constants. It enables this:

```python
full_name = n_valid_chars_from(full_name, n=max_chars_from_full_name)
full_name = full_name or "Anonymous"

user_id = n_valid_chars_from(user_id, n=max_chars_from_user_id)
user_id = pad(user_id, num_chars_from_user_id)

return full_name + user_id
```

(I've changed the function and parameter names to shorter and hopefully nicer ones too.)

**But** these little functions are really just part of the `generate_username()` function and not meant for reuse, as you point out, so I've made them internal functions within `generate_username()`.